Sources/SwiftRefactor/ExpandEditorPlaceholder.swift

```diff
@@ -0,0 +1,347 @@
+/// call, expands it and any adjacent function-typed placeholders to trailing
+/// closures on that call. All other placeholders will expand as per
+/// `ExpandEditorPlaceholder`.
+///
+/// ## Before
+/// ```swift
+/// foo(closure1: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, normalArg: <#T##Int#>, closure2: { ... }, closure3: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, closure4: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>
+/// ```
+///
+/// ## `closure3` or `closure4` Expansion
+/// ```swift
+/// foo(closure1: <#T##(Int) -> String##(Int) -> String##(_ someInt: Int) -> String#>, normalArg: <#T##Int#>, closure2: { ... }) { someInt in
+///   <#T##String#>
+/// } closure2: { someInt in
+///   <#T##String#>
+/// }
+/// ```
+///
+/// Expansion on `closure1` and `normalArg` is the same as `ExpandEditorPlaceholder`.
+public struct ExpandEditorPlaceholders: RefactoringProvider {
+  public static func refactor(syntax token: TokenSyntax, in context: Void) -> ExprSyntax? {
+    guard let placeholder = token.parent?.as(EditorPlaceholderExprSyntax.self),
+      let arg = placeholder.parent?.as(TupleExprElementSyntax.self),
+      let argList = arg.parent?.as(TupleExprElementListSyntax.self),
+      let call = argList.parent?.as(FunctionCallExprSyntax.self)
+    else {
+      return ExpandEditorPlaceholder.refactor(syntax: token)
+    }
+
+    guard let expanded = call.expandTrailingClosurePlaceholders(ifIncluded: arg) else {
+      return ExpandEditorPlaceholder.refactor(syntax: token)
+    }
+
+    guard let trailing = CallToTrailingClosures.refactor(syntax: expanded.expr, in: .init(startAt: argList.count - expanded.numClosures)) else {
+      return ExpandEditorPlaceholder.refactor(syntax: token)
+    }
+
+    return ExprSyntax(trailing)
+  }
+}
+
+extension FunctionTypeSyntax {
+  /// Return a closure expression for this function type, eg.
+  /// ```
+  ///   (_ someInt: Int) -> String
+  /// ```
+  /// would become
+  /// ```
+  /// { someInt in
+  ///   <#T##String#>
+  /// }
+  /// ```
+  fileprivate var closureExpansion: ClosureExprSyntax {
+    let closureSignature: ClosureSignatureSyntax?
+    if !arguments.isEmpty {
+      var args = arguments.dropLast().map { ClosureParamSyntax(name: $0.expansionNameToken(), trailingComma: .commaToken()) }
+      args.append(ClosureParamSyntax(name: arguments.last!.expansionNameToken()))
+      closureSignature = ClosureSignatureSyntax(input: .simpleInput(ClosureParamListSyntax(args)))
+    } else {
+      closureSignature = nil
+    }
+
+    // Single statement for the body - the placeholder-ed type if non-void and
+    // 'code' otherwise.
+    let ret = output.returnType.description
+    let placeholder: String
+    if ret == "Void" || ret == "()" {
+      placeholder = ExpandEditorPlaceholder.wrapInTypePlaceholder("code", type: "Void")
+    } else {
+      placeholder = ExpandEditorPlaceholder.wrapInTypePlaceholder(ret, type: ret)
+    }
```

Comment:
Do we even need the `if` here? The only time it does anything is translate `()` to `Void` and I would imagine that all parts of the compiler that deal with placeholders know that `()` and `Void` are equivalent.

Reply:
The main purpose isn't for making `()` `Void`, it's to make `Void` show `code` and not `Void` which IMO is a nice little change 😅.

Reply:
Ah, I missed the `"code"` vs `ret` distinction. Makes sense 👍🏽